### PR TITLE
18 change energy move

### PR DESCRIPTION
**Pull request by gfabbris**
This is related to my attempts to make the mono, undulatory and phase retarders track their energies in a cleaner way, see #18. 

The ultimate goal is to be able to move all these energy-related devices using something like `mv(mono.energy, 10)`, i.e. without the need of a specific `moveE` function. While this is not achieved in this pull request, I've managed to clean up `moveE` quite a bit by moving some of its contents to the devices, in particular:

- The phase retarders now have a `.energy` `PseudoPositioner` that is coupled to `.th`. For instance, to move the pr1 to 10 keV we can just do `mv(pr1.energy, 10)`.
- The undulator backlash is now handled in the device.

It also includes some fixes and improvements to `pr_setup`.

### CODE REVIEW

**Comment by gfabbris** [comment]
In this last version I created an `energy` `Signal` that will correspond to the *beamline energy*. Here, I am explicitly forcing `energy.get() = mono.energy.get()`, i.e., the monochromator is our energy reference, but it doesn't need to be like this for other beamlines. Moving the `energy` device will check and move whatever energy-related devices are tracking.

**Comment by strempfer** [comment]
At PETRA III, we had Tango devices called 'multiplemotor' which could be used to move several motors at the same time and get a done signal only after the last motor was done moving. This we used also to define slave devices to the monochromator. Perhaps we should also do this rather on the EPICS level. Here is the corresponding Python code for setting this in Sardana to give you an idea how it worked:

class eslave(Macro):
    """
        eslave - Defines slave devices to the monochromator energy 
    """
    param_def = [
        [ "number", Type.Float, -999, "Slave devices" ],
        ]
    interactive=True
    def run( self, number):   
        self.multiple_motors = PyTango.DeviceProxy("haspp09mono:10000/p09/multiplemotors/mono.01")
        if number != -999:
            self.multiple_motors.MotorMask = number
            self.output( "Multiple Motors set to %4.0f " % number)
        else:
            number = self.multiple_motors.MotorMask
            self.output( "Possibilities:  no slave device  (0)")
            self.output( "                undulator        (1)")
            self.output( "                analyzer EH1     (2)")
            self.output( "                analyzer EH2     (4)")
            self.output( "                attenfactor EH1  (8)")
            self.output( "                attenfactor EH2 (16)")
            self.output( "                PR1             (32)")
            self.output( "                PR2             (64)")
            self.output( "                PR3            (128)")
            self.output( "                PR4            (256)")
            self.output( "")
            number=self.input( "New setting?" ,default_value=number)
            self.multiple_motors.MotorMask = int(number)
            self.output( "Multiple Motors set to %s " % number)

**Comment by gfabbris** [comment]
@strempfer - As far as I am aware, Bluesky doesn't have something equivalent.

I tested the last commit (d67f34f) at the beamline, and it works nicely. But I realized one remaining issue, when we move `energy` only its value is recorded in the event, which here is the monochromator energy. We would like to also record the undulator and PRs positions, but I don't know yet how to do it. My first idea was to include these in the `energy.read()`, but this will generate duplicate entries in the database (for instance in the baseline stream), so it will likely fail. Another way would be to add some logic to our local scans to add the other devices to the `detectors` list when `energy` is a positioner, but it'd be better to have a solution that would also work with `bluesky.plans`.

**Comment by prjemian** [comment]
Bluesky *can* move several motors at once and wait for the batch of them to all complete before proceeding.

```python
yield from bps.mv(
    m1, 5,
    m2, 10,
    m3, 15,
    m41, 2.75,
)
```

The `multiplemotor` concept could be implemented in ophyd as a [Device](https://blueskyproject.io/ophyd/device-overview.html#device) or a [PVPositioner](https://blueskyproject.io/ophyd/positioners.html#pvpositioner)

**Comment by gfabbris** [comment]
For now, I left the *energy.py* file in place, but it will not load it (i.e. no `energy` device). But in any case, I think its time to review and merge this into the main branch, because there are some good improvements here. Mainly, a lot of the motion of the undulator and phase retarders are now handled in the devices, which makes the `moveE` plan much simpler.

**Comment by gfabbris** [comment]
I think this looks good now. Main changes:

- Adds the `energy_scan_decorator`: if an energy scan is performed, the relevant PR and/or undulator is added to the detector list (so that their position is read). This decorator changes the hinting of these devices so that the BestEffortCallback won't plot them.
- Creates a new the `qxscan` plan that is based on the `bluesky.plans.list_scan`. The detectors and the counting time per point are passed as part of the `*args`, and are moved using `det.preset_monitor`.

#### With these changes, the dedicated `energy_plans.py` becomes obsolete.
#### There is substantial overlap between this pull request and #79, so will need to be careful when merging.

File: profile_bluesky/startup/instrument/plans/local_scans.py [hunk 1]
```diff
@@ -157,15 +188,69 @@ def ascan(*args, monitor=None, detectors=[scalerd], lockin=False,
...
+        else:
+            for det in detectors:
+                _ct[det] = abs(monitor)
+                args += (det.preset_monitor, abs(monitor)*array(factor_list))
+    else:
+        for det in detectors:
+            _ct[det] = yield from rd(det.preset_monitor)
+            args += (det.preset_monitor, _ct[det]*array(factor_list))
+
+    @energy_scan_decorator(True)
```
Comment (gfabbris) [inline profile_bluesky/startup/instrument/plans/local_scans.py:240]: This needs to be: `@energy_scan_decorator(True, extras)`

**Pull request merged**